larq/optimizers_v1.py

```diff
@@ -0,0 +1,180 @@
+            raise NotImplementedError(
+                "Xavier Learning rate scaling not implimented for this kernelsize"
+            )
+        return coeff
+
+    def get_updates(self, loss, params):
+        mult_lr_params = [p for p in params if p.name in self.multipliers]
+        base_lr_params = [p for p in params if p.name not in self.multipliers]
+
+        updates = []
+        base_lr = self.optimizer.lr
+        for param in mult_lr_params:
+            self.optimizer.lr = base_lr * self.multipliers[param.name]
+            updates.extend(self.optimizer.get_updates(loss, [param]))
+
+        self.optimizer.lr = base_lr
+        updates.extend(self.optimizer.get_updates(loss, base_lr_params))
+
+        return updates
+
+    def __getattr__(self, name):
+        return getattr(self.optimizer, name)
+
+    def get_config(self):
+        return {
+            "optimizer": {
+                "class_name": self.optimizer.__class__.__name__,
+                "config": self.optimizer.get_config(),
+            },
+            "multipliers": self.multipliers,
+        }
+
+    @classmethod
+    def from_config(cls, config, custom_objects=None):
+        optimizer = tf.keras.optimizers.deserialize(
+            config.pop("optimizer"), custom_objects=custom_objects
+        )
+        return cls(optimizer, config["multipliers"])
+
+
+@utils.register_keras_custom_object
+class Bop(tf.keras.optimizers.Optimizer):
+    def __init__(self, fp_optimizer, threshold=1e-5, gamma=1e-2, name="Bop", **kwargs):
+        super().__init__(**kwargs)
+
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
+            )
+
+        with tf.keras.backend.name_scope(self.__class__.__name__):
+
+            self.fp_optimizer = fp_optimizer
+            self.threshold = threshold
+            self.gamma = gamma
+            self.iterations = tf.keras.backend.variable(
+                0, dtype="int64", name="iterations"
+            )
+
+    def get_updates(self, loss, params):
+        grads = self.get_gradients(loss, params)
+        self.updates = []
+
+        ms = [
+            tf.keras.backend.zeros(
+                tf.keras.backend.int_shape(p), dtype=tf.keras.backend.dtype(p)
+            )
+            for p in params
+        ]
+
+        fp_params = []
+
+        for p, g, m in zip(params, grads, ms):
+            if "/kernel" in p.name and "quant_" in p.name:
+                m_t = (1 - self.gamma) * m + self.gamma * g
+
+                self.updates.append(tf.assign(m, m_t))
+                self.updates.append(
+                    tf.assign(
+                        p, lq.quantizers.sign(-p * tf.sign(p * m_t - self.threshold))
+                    )
+                )
+
+            else:
+                fp_params.append(p)
+
+        self.updates.append(tf.assign(self.iterations, self.iterations + 1))
```

Comment:
Either remove, or use
```suggestion
        self.updates.append(tf.assign_add(self.iterations, 1))
```